types/index.d.ts

```diff
@@ -31,14 +51,14 @@ export interface RouteComponentProps<T extends DefaultParams = DefaultParams> {
   params: T;
 }
 
-export interface RouteProps<T extends DefaultParams = DefaultParams> {
-  children?: ((params: Params<T>) => ReactNode) | ReactNode;
-  path?: Path;
-  component?: ComponentType<RouteComponentProps<T>>;
+export interface RouteProps<RoutePath extends Path = Path> {
+  children?: ((params: ExtractRouteParams<RoutePath>) => ReactNode) | ReactNode;
+  path?: RoutePath;
+  component?: ComponentType<RouteComponentProps<ExtractRouteParams<RoutePath>>>;
```

Comment:
I was wondering if there is any way to keep the backward compatibility for users who want to manually type their params, like:

```typescript
Route<CustomParams> // params is of type `CustomParams`
Route // the type of params is inferred from path
```

Maybe we could move `RoutePath` to the second type argument and provide a default first argument like so:

```typescript
export function Route<T extends DefaultParams = null, RoutePath extends Path = Path>(
```
But I have a little experience, so not sure if this will work.